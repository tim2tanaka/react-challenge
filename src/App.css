.App {
  text-align: center;
}

.app-header {
  background-color: #282c34;
  display: flex;
  flex-direction: column;
  font-size: 1.5rem;
  color: white;
  padding: 8px;
  margin: 8px;
  align-items: center;
}

.dropdownList {
  display: flex;
  flex-direction: column;
  font-size: 1rem;
  padding: 8px;
  margin: 8px;
  align-items: center;
  justify-content: center;
}

.dropdownList-hide {
  display: none;
  visibility: hidden;
}

.select-dropdown {
  font-size: 1rem;
}

.dashed-line-boarder {
  padding: 8px 0px 0px 0px;
  margin: 8px;
  border-top: 3px dashed #000000;
}

.pagination {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content:space-around;
  padding: 8px;
  margin: 8px;
}

.pagination-hide {
  display: none;
  visibility: hidden;
}

.search-input {
  padding: 8px;
  margin: 8px;
  border: 1px solid black
}

.search {
  padding: 8px;
  margin: 8px;
}

.button {
  display: inline-block;
  padding: 8px;
  margin: 8px;
  border: 1px solid black;
  cursor: pointer;
}

.search-display {
  padding: 8px;
  margin: 8px;
}

.error-msg {
  background-color: #ffffff;
  display: flex;
  flex-direction: column;
  font-size: 1rem;
  font-family: 'Courier New', Courier, monospace;
  color: rgb(255, 0, 0);
  padding: 8px;
  margin: 8px;
  align-items: center;
}

.loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 8px;
  margin: 8px;
}

.loading-hide {
  display: none;
  visibility: hidden;
}

.loading-logo {
  height: 15vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .loading-logo {
    animation: loading-logo-spin infinite 1s linear;
  }
}

@keyframes loading-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}