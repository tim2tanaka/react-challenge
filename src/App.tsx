import './App.css';
import { Header } from './components/Header';
import { Search } from './components/Search';
import { SearchDisplay } from './components/SearchDisplay';
import { ErrorDisplay } from './components/ErrorDisplay';
import { Pagination } from './components/Pagination';
import { DropdownList } from './components/PagesDropDown';
import { Loading } from './components/Loading';
import { useAppState, useAppDispatch } from './app_state/context';
import { actions } from './app_state/actions';
import { Data } from './types';

function App() {
  const state = useAppState();
  const dispatch = useAppDispatch();

  async function loadData(searchText:string, pageNumber: number) {
    const searchTxt = searchText ? searchText : state.search;
    const currentPage = pageNumber ? pageNumber : state.pagination.page;
    dispatch({ type: actions.setIsloading.type });

    function getPaginationData(data: Data['data']) {
      const { resultsPerPage } = state.pagination;
      const resultsPage = currentPage <= 0 ? 1 : currentPage;
      const startIndex = (resultsPage - 1) * resultsPerPage;
      const endIndex = startIndex + resultsPerPage;
      dispatch({ type: actions.setTotalPages.type, payload: data.length });
      return data.slice(startIndex, endIndex);
    }

    const url = 'https://restcountries.com/v3.1/name';
    try {
      dispatch({ type: actions.clearData.type, payload: [] });
      dispatch({ type: actions.clearErrorMsg.type });
      const res = await fetch(`${url}/${searchTxt}`);
      if (res.status !== 200) throw new Error('Country search failed!');
      const data = await res.json();
      const paginationData: Data['data'] = getPaginationData(data);
      dispatch({ type: actions.setData.type, payload: paginationData });
      dispatch({ type: actions.setIsloading.type });
      dispatch({ type: actions.setTotalPages.type, payload: data.length });
    } catch (error: any) {
      if (error?.message)
        dispatch({ type: actions.setErrorMsg.type, payload: error.message });
      dispatch({ type: actions.setIsloading.type });
    }
  }

  return (
    <>
        <Header title="Country Search App" />
        <Search loadData={loadData}/>
        <DropdownList loadData={loadData} />
        {!state.isloading && <SearchDisplay />}
        <ErrorDisplay />
        {state.isloading && <Loading />}
        <Pagination loadData={loadData} />
    </>
  );
}

export default App;
