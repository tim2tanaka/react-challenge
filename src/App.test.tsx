import { render, screen } from '@testing-library/react';
import App from './App';

describe ('App', ()=> {
  test('render App', () => {
    const { container } = render(<App />)
    expect(container).toMatchSnapshot()

  });
  test('renders header', () => {
    render(<App />);
    const headerElement = screen.getByText(/Country search/i);
    expect(headerElement).toBeInTheDocument();
  });
});
