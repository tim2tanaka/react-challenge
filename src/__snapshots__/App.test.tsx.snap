// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render App 1`] = `
<div>
  <header
    class="App-header"
  >
    <h2>
      Country search
    </h2>
  </header>
  <div
    class="Search"
  >
    <input
      class="searchInput"
      placeholder="Search..."
      type="text"
    />
    <button
      class="button"
    >
      Search
    </button>
  </div>
  <div
    id="search-results-container"
  />
</div>
`;
